hydro_cli/src/lib.rs

```diff
@@ -20,17 +21,69 @@ pub mod core;
 
 create_exception!(hydro_cli_core, AnyhowError, PyException);
 
+#[pyclass]
+struct SafeCancelToken {
+    cancel_tx: Option<Sender<()>>,
+}
+
+#[pymethods]
+impl SafeCancelToken {
+    fn safe_cancel(&mut self) {
+        if let Some(token) = self.cancel_tx.take() {
+            eprintln!("Received cancellation, cleaning up...");
+            token.send(()).unwrap();
+        } else {
+            eprintln!("Already received cancellation, please be patient!");
+        }
+    }
+}
```

Comment:
Different from https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html ?

Reply:
It just needs to be exposed to Python, which is why we wrap it